### ISSUE TITLE
Mounting volumes for persistence

### ISSUE THREAD

**Issue by pogzie** [wwmoraes/pvpgn-server-docker#3]
Im trying to use this so that I can have a mounted volume for data and config persistence but i cant seem to get it to work. Any tips? Also the table of contents in the readme has something about deployment but that section doesnt exist.

```
docker run --name docker-pvpgn -d  -v /root/docker-pvpgn/etc:/usr/local/pvpgn/etc -v /root/docker-pvpgn/var:/usr/local/pvpgn/var -p 4000:4000 -p 6112:6112 wwmoraes/pvpgn-server
```

**Comment by wwmoraes** [comment]
Thanks for the heads up about the readme! I'll check it and your volume problem tomorrow, then get back to you ;)

**Comment by pogzie** [comment]
To add, im logged in as `root` and the `/root/docker-pvpgn/etc` (and var) exists.

**Comment by wwmoraes** [comment]
I assume those folders are empty right? When you mount a volume/bind mount, the files inside the container are ignored. To solve that, you need to copy all the configuration and state files into your local folder or into a docker volume. I've added instructions on the readme under the deployment section, and a script to simplify creating such volumes.

TL;DR in your case do

```shell
# copy var files into the host folder
docker run --rm -v /root/docker-pvpgn/var:/mnt/var wwmoraes/pvpgn-server cp -r /usr/local/pvpgn/var/pvpgn /mnt/var
# copy etc files into the host folder
docker run --rm -v /root/docker-pvpgn/etc:/mnt/etc wwmoraes/pvpgn-server cp -r /usr/local/pvpgn/etc/pvpgn /mnt/etc
# run it
docker run --name docker-pvpgn -d  -v /root/docker-pvpgn/etc:/usr/local/pvpgn/etc -v /root/docker-pvpgn/var:/usr/local/pvpgn/var -p 4000:4000 -p 6112:6112 wwmoraes/pvpgn-server
```

**Comment by wwmoraes** [comment]
@pogzie does that solve your problem?

**Comment by ibepeer** [comment]
hiya! I have the same problem. 
I am getting permission denied

```
root@server:~/docker-pvpgn# docker run --rm -v /root/docker-pvpgn/var:/mnt/var wwmoraes/pvpgn-server cp -r /usr/local/pvpgn/var/pvpgn /mnt/var 
cp: can't create directory '/mnt/var/pvpgn': Permission denied
```

**Comment by wwmoraes** [comment]
@ibepeer this is not an error on the container, but rather on the host - check if you have [user mapping](https://docs.docker.com/engine/security/userns-remap/) in place, as it might mess up permissions. Also try with a fresh local var folder, e.g.:

```shell
mkdir var && docker run --rm -v $PWD/var:/mnt/var wwmoraes/pvpgn-server cp -r /usr/local/pvpgn/var/pvpgn /mnt/var
```

**Comment by ibepeer** [comment]
@wwmoraes strange. I dont have any usermappings in place, also deleted the folder and tried to run the command again to no avail.

I managed to get `make run` and `make sh` to work. I went ahead and modified the Dockerfile to use the newest `develop` branch as suggested by a dev.

Thanks for looking into this and it seems to be running with make run/make sh.

```
FROM alpine:latest as builder

ARG REPO=https://github.com/pvpgn/pvpgn-server.git
ARG WITH_MYSQL=true
ARG WITH_LUA=true

### Install build dependencies
RUN apk --no-cache add git build-base clang curl-dev libcurl cmake make zlib-dev lua-dev mariadb-dev

### CMake & make
RUN git clone --single-branch --branch develop ${REPO} /src
RUN mkdir build
WORKDIR /src/build
RUN mkdir /usr/local/pvpgn
RUN cmake \
  -D WITH_MYSQL=${WITH_MYSQL} \
  -D WITH_LUA=${WITH_LUA} \
  -D CMAKE_INSTALL_PREFIX=/usr/local/pvpgn \
  -w \
  ../

### Install
RUN make
RUN make install

FROM alpine:latest as runner

### Install dependencies
RUN apk --no-cache add ca-certificates libstdc++ libgcc libcurl curl-dev lua5.1-libs mariadb-connector-c

### Copy build files
COPY --from=builder /usr/local/pvpgn /usr/local/pvpgn

### Prepare user
RUN addgroup --gid 1001 pvpgn \
  && adduser \
  --home /usr/local/pvpgn \
  --gecos "" \
  --shell /bin/false \
  --ingroup pvpgn \
  --system \
  --disabled-password \
  --no-create-home \
  --uid 1001 \
  pvpgn

### Make volume folders
RUN mkdir -p /usr/local/pvpgn/var
RUN mkdir -p /usr/local/pvpgn/etc
RUN chown -R pvpgn:pvpgn /usr/local/pvpgn/var
RUN chown -R pvpgn:pvpgn /usr/local/pvpgn/etc

### persist data and configs
VOLUME /usr/local/pvpgn/var
VOLUME /usr/local/pvpgn/etc

### adjust permissions
RUN chown -R pvpgn:pvpgn /usr/local/pvpgn

# expose served network ports
EXPOSE 6112 4000

### Set user
USER pvpgn

### RUN!
CMD ["/usr/local/pvpgn/sbin/bnetd", "-f"]
```

**Comment by wwmoraes** [comment]
@ibepeer I've added an extra build argument, `BRANCH`, which allows you to override the source branch used during clone. I also added a make target, `build-develop`, to make it easier to build a bleeding-edge version from that branch. I'll keep the image on docker hub and the latest tag based off master though, to prevent breaking anything for current users.
